### PR TITLE
Fix hostpath pv recycle bug

### PR DESCRIPTION
**Pull request by llhhbc**
1. Fix pv recycle share one pod object
     2. Fix pv recycle pod run on unexpected node

**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
I have create many pv with hostpath and persistentVolumeReclaimPolicy: Recycle
There is bugs:
1. After delete pvc, the pv may be status of Failed.
2. After delete pvc, hostpath may not be cleaned.

Here is logs:

```log
k get pv
NAME             CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS      CLAIM                                                STORAGECLASS   REASON   AGE
pv-hostpath-19   10Gi       RWO            Recycle          Failed      default/example-vttablet-zone1-1996462234-c7ea0d33                           20h

k get po
NAME                               READY   STATUS      RESTARTS   AGE
recycler-for-pv-hostpath-15        0/1     Completed   0          2m12s

k describe pv pv-hostpath-19
Name:              pv-hostpath-19
Labels:            type=local
Annotations:       pv.kubernetes.io/bound-by-controller: yes
Finalizers:        [kubernetes.io/pv-protection]
StorageClass:
Status:            Failed
Claim:             default/example-vttablet-zone1-1996462234-c7ea0d33
Reclaim Policy:    Recycle
Access Modes:      RWO
VolumeMode:        Filesystem
Capacity:          10Gi
Node Affinity:
  Required Terms:
    Term 0:        kubernetes.io/hostname in [10-10-30-184]
Message:           Recycle failed: failed to delete recycler pod: pods "recycler-for-pv-hostpath-26" not found
Source:
    Type:          HostPath (bare host directory volume)
    Path:          /root/pv/pv19
    HostPathType:
Events:
  Type     Reason               Age                From                         Message
  ----     ------               ----               ----                         -------
  Warning  VolumeFailedRecycle  91s                persistentvolume-controller  Recycle failed: failed to delete recycler pod: pods "recycler-for-pv-hostpath-26" not found
  Normal   RecyclerPod          86s (x2 over 93s)  persistentvolume-controller  Recycler pod: Successfully assigned default/recycler-for-pv-hostpath-19 to 10-10-30-184
  Normal   RecyclerPod          86s (x2 over 92s)  persistentvolume-controller  Recycler pod: Container image "busybox:1.27" already present on machine
  Normal   RecyclerPod          86s (x2 over 92s)  persistentvolume-controller  Recycler pod: Created container pv-recycler
  Normal   RecyclerPod          86s (x2 over 92s)  persistentvolume-controller  Recycler pod: Started container pv-recycler
```

**Which issue(s) this PR fixes**:
None

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/volume/hostpath/host_path.go [hunk 1]
```diff
@@ -151,6 +151,15 @@ func (plugin *hostPathPlugin) Recycle(pvName string, spec *volume.Spec, eventRec
 			Path: spec.PersistentVolume.Spec.HostPath.Path,
 		},
 	}
+	if spec.PersistentVolume.Spec.NodeAffinity != nil {
```
Comment (msau42) [inline pkg/volume/hostpath/host_path.go:154]: add unit tests for these?

**Review by msau42 — COMMENTED** [review]
BTW, the recycling reclaim policy has been [deprecated](https://github.com/kubernetes/kubernetes/issues/59060) since 1.10. Any reason why the [sig-storage-local-static-provisioner](https://github.com/kubernetes-sigs/sig-storage-local-static-provisioner) recycling-like behavior does not work?

File: pkg/volume/hostpath/host_path.go [hunk 2]
```diff
@@ -142,7 +142,7 @@ func (plugin *hostPathPlugin) Recycle(pvName string, spec *volume.Spec, eventRec
 		return fmt.Errorf("spec.PersistentVolume.Spec.HostPath is nil")
 	}
 
-	pod := plugin.config.RecyclerPodTemplate
+	pod := plugin.config.RecyclerPodTemplate.DeepCopy()
```
Comment (msau42) [inline pkg/volume/hostpath/host_path.go:145]: good catch!

Comment (msau42) [reply pkg/volume/hostpath/host_path.go:145]: Would it be worth adding a unit test that recycles twice to catch this error?

File: pkg/volume/hostpath/host_path.go (hunk 1 above)
Comment (llhhbc) [reply pkg/volume/hostpath/host_path.go:154]: This is just for prevent nil panic, hostPath localpv must have NodeAffinity, will not be nil.

File: pkg/volume/hostpath/host_path.go (hunk 2 above)
Comment (llhhbc) [reply pkg/volume/hostpath/host_path.go:145]: I will try to add unit test.

Comment (llhhbc) [reply pkg/volume/hostpath/host_path.go:145]: Is hard to add unit test for fake client. I have test in real kubernetes environment.

**Comment by llhhbc** [comment]
@msau42 Can this to be merge?

**Comment by cbandy** [comment]
@llhhbc did you see kublet logs similar to those in #92946? I'm wondering if this change also fixes that (stale) issue.

**Comment by llhhbc** [comment]
> @llhhbc did you see kublet logs similar to those in #92946? I'm wondering if this change also fixes that (stale) issue.

yes，I think so.

File: pkg/volume/hostpath/host_path.go (hunk 2 above)
Comment (msau42) [reply pkg/volume/hostpath/host_path.go:145]: Sorry for the delay. What do you think of splitting this function into two pieces: 1) One that generates the recycler pod spec, and 2) one that invokes the pod. Then we can unit test the first.

**Pull request closed without merging**